### ISSUE TITLE
Incorrect error message text for instantiating abstract class

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#25283]
|  |  |
| --- | --- |
| Bugzilla Link | [24909](https://llvm.org/bz24909) |
| Version | unspecified |
| OS | All |
| Attachments | [Proposed patch](https://user-images.githubusercontent.com/60944935/143752056-c6dbaf73-10ac-414e-b1df-bbce56ca4e80.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor |

## Extended Description 
When attempting to instantiate an abstract class clang returns:

test.cc:18:5: error: variable type 'A' is an abstract class
  A a;
    ^
test.cc:4:16: note: unimplemented pure virtual method 'a' in 'A'
  virtual void a()=0;

The second line claims that A::a() is unimplemented.  It does this even if you include an implementation.  This is incorrect and potentially misleading.

```cpp
struct A {
  virtual void a()=0;
};

void A::a() {}

int main() {
  A a;
}
```

**Comment by shafik** [comment]
gcc/edg/MSVC: https://godbolt.org/z/eqYfaz9hx

Seem to provide a easier to understand diagnostic, I think it may be worth considering updating our diagnostic.

CC @AaronBallman
